src/lib/FlightTasks/tasks/ManualAcceleration/FlightTaskManualAcceleration.cpp

```diff
@@ -0,0 +1,151 @@
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file FlightTaskManualAcceleration.cpp
+ */
+
+#include "FlightTaskManualAcceleration.hpp"
+#include <mathlib/mathlib.h>
+#include <float.h>
+
+using namespace matrix;
+
+bool FlightTaskManualAcceleration::activate(vehicle_local_position_setpoint_s last_setpoint)
+{
+	bool ret = FlightTaskManual::activate(last_setpoint);
+	_velocity_setpoint.zero();
+	return ret;
+}
+
+bool FlightTaskManualAcceleration::update()
+{
+	// maximum commanded acceleration and velocity
+	Vector3f acceleration_scale;
+	Vector3f velocity_scale;
+	acceleration_scale(0) = acceleration_scale(1) = _param_mpc_acc_hor.get();
+	velocity_scale(0) = velocity_scale(1) = _param_mpc_vel_manual.get();
+
+	if (_velocity(2) < 0.f) {
+		acceleration_scale(2) = _param_mpc_acc_up_max.get();
+		velocity_scale(2) = _param_mpc_z_vel_max_up.get();
+
+	} else {
+		acceleration_scale(2) = _param_mpc_acc_down_max.get();
+		velocity_scale(2) = _param_mpc_z_vel_max_dn.get();
+	}
+
+	acceleration_scale *= 2.f; // because of drag the average aceleration is half
+
+	// Yaw
+	_position_lock.updateYawFromStick(_yawspeed_setpoint, _yaw_setpoint,
+					  _sticks_expo(3) * math::radians(_param_mpc_man_y_max.get()), _yaw, _deltatime);
+
+	// Map stick input to acceleration
+	Vector2f stick_xy(&_sticks_expo(0));
+	_position_lock.limitStickUnitLengthXY(stick_xy);
+	_position_lock.rotateIntoHeadingFrameXY(stick_xy, _yaw, _yaw_setpoint);
+
+	_acceleration_setpoint = Vector3f(stick_xy(0), stick_xy(1), _sticks_expo(2)).emult(acceleration_scale);
+
+	// Add drag to limit speed and brake again
+	Vector3f cv = acceleration_scale.edivide(velocity_scale);
+	_acceleration_setpoint -= cv.emult(_velocity);
+
+	lockAltitude();
+	lockPosition(stick_xy.length());
+
+	_constraints.want_takeoff = _checkTakeoff();
+	return true;
+}
+
+void FlightTaskManualAcceleration::lockAltitude()
+{
+	// integrate vertical velocity
+	const float velocity_setpoint_z_last = _velocity_setpoint(2);
+	_velocity_setpoint(2) += _acceleration_setpoint(2) * _deltatime;
+
+	if (fabsf(_sticks_expo(2)) > FLT_EPSILON) {
+		_position_setpoint(2) = NAN;
+
+	} else {
+		if (!PX4_ISFINITE(_position_setpoint(2))) {
+			_position_setpoint(2) = _position(2);
+		}
+
+		_position_setpoint(2) += _velocity_setpoint(2) * _deltatime;
+
+		if (fabsf(_velocity_setpoint(2)) > fabsf(velocity_setpoint_z_last)) {
+			_velocity_setpoint(2) = velocity_setpoint_z_last;
+		}
+	}
+}
+
+void FlightTaskManualAcceleration::lockPosition(const float stick_input_xy)
+{
+	if (stick_input_xy > FLT_EPSILON) {
+		_position_setpoint(0) = _position_setpoint(1) = NAN;
```

Comment:
`_position_setpoint.xy() = Vector2f(NAN, NAN)` ? or does it create too much overhead?

Reply:
or create the `position_xy` and `velocity_xy` before the `if`, set it to NAN here and move the `position_xy -> _position_setpoint` of lines 138-141 outside of the `if`